### ISSUE TITLE
Specified namespace is incorrectly returned in response when eval'ing in non-default namespace

### ISSUE THREAD

**Issue by cursive-ide** [nrepl/nrepl#171]
When `eval`ing forms in a non-default namespace, the selected namespace is incorrectly returned as part of the `:value` message, leading Cursive to believe that `*ns*` has changed. Here's a sample transcript:

Here I eval `*ns*` by typing it into the REPL, so no `:ns` is sent. The current REPL ns is indeed `user`.

```
{:op "eval", 
 :code "*ns*", 
 :session "a167fcd8-61d2-4be8-9715-98baf9ead05b", 
 :id "acbaf8f0-dbf6-45d3-9cdd-15dec095d17a"}
```

```
{:id "acbaf8f0-dbf6-45d3-9cdd-15dec095d17a", 
 :ns "user", 
 :session "a167fcd8-61d2-4be8-9715-98baf9ead05b", 
 :value "#object[clojure.lang.Namespace 0x24411926 \"user\"]"}
```

Here I send a form `(foo)` from a file open in an editor containing the `dev.scratch` namespace. Here Cursive sends the `:ns` field, and note that the same namespace is incorrectly returned in the `:ns` field on the reply.

```
{:ns "dev.scratch", 
 :op "eval", 
 :code "^{:clojure.core/eval-file \"/Users/colin/dev/cursive-bugs/namespace-switch-problem/src/dev/scratch.clj\" :line 7 :column 3} (foo)", 
 :session "a167fcd8-61d2-4be8-9715-98baf9ead05b", 
 :id "5209f36c-a36e-434a-9d11-a22197c7a73d"}
```

```
{:id "5209f36c-a36e-434a-9d11-a22197c7a73d", 
 :ns "dev.scratch", 
 :session "a167fcd8-61d2-4be8-9715-98baf9ead05b", 
 :value "nil"}
```

Eval'ing `*ns*` again shows that the current namespace has not actually changed.

```
{:op "eval", 
 :code "*ns*", 
 :session "a167fcd8-61d2-4be8-9715-98baf9ead05b", 
 :id "075deef1-21fb-44b1-a911-70ae896963ef"}
```

```
{:id "075deef1-21fb-44b1-a911-70ae896963ef", 
 :ns "user", 
 :session "a167fcd8-61d2-4be8-9715-98baf9ead05b", 
 :value "#object[clojure.lang.Namespace 0x24411926 \"user\"]"}
```

Clojure 1.10.1
Java 11
macOS Mojave

**Comment by shen-tian** [comment]
Is your expected behaviour that `*ns*` should be `dev.scratch` in the last response? 

Looking at what docs are around, it is a bit ambiguous what `:ns` in response means, but as far as I can tell, it's the last value of `*ns*` after eval **but before restoring** the ns to the original. Guess this should be same as provided `:ns`, unless part of eval also changes the namespace. 

Looking at the source, the restoration of ns after eval definitely seems intentional.

Does this behaviour cause any problems on your side?

**Comment by cursive-ide** [comment]
No, my expected behaviour is either that the value returned in the `:ns` field on the evaluation result of `(foo)` should be the actual current REPL namespace, i.e. `user`, or it should be absent if the command didn't actually affect `*ns*`. 

Currently Cursive assumes that whatever comes back in a response message is the current ns, and that always used to work correctly. I don't know if this case previously didn't return `:ns` at all, or if it previously used to return the current value of `*ns*`. I'll see if I can test that today. 

I need the current ns to display it to the user in the UI since Cursive doesn't display the standard REPL prompt. If I can't rely on this then I'll have to do another round-trip after every evaluation to get the value of `*ns*`, which would suck.

**Comment by shen-tian** [comment]
> ... and that always used to work correctly.

Would be handy if you can check in which version this behaviour changed?

**Comment by cursive-ide** [comment]
Yes, I spent some time yesterday digging through the code, and I can't see any way that previous versions of the code would have worked as I recall them working. I'm going to write a small project to try to reproduce it and I'll try it with various versions to see what happens.

**Comment by shen-tian** [comment]
@cursive-ide : any luck reproducing this?

**Comment by cursive-ide** [comment]
@shen-tian No sorry, I got distracted with other things. I'll try to repro this today.

**Comment by cmf** [comment]
Ok, so I must have been deluded with this, I've tested this back to org.clojure/tools.nrepl 0.2.12 and none of the versions behave as I remembered. It's very strange because I'm sure Cursive's behaviour has changed (and other users have noticed this too, and filed issues). However I have no plausible explanation for _how_ it used to work. Cursive has always just taken any `:ns` field from the reply messages and shown it to the user.

One question is, how _should_ it work? As far as I can tell, the correct logic should be to just ignore the `:ns` response to messages with an `:ns` in the request, since you can't actually influence `*ns*` from them anyway:

```clojure
(doall (nrepl/message (nrepl/client conn 1000) {:op :eval :ns "test" :code "(ns test2)"}))
=>
({:id "553c7e0f-1d2f-47c7-acb5-58c6c838f7f4",
  :ns "test2",
  :session "bdeacbba-6fe9-482e-b2f3-cc81114bfff3",
  :value "nil"}
 {:id "553c7e0f-1d2f-47c7-acb5-58c6c838f7f4", :session "bdeacbba-6fe9-482e-b2f3-cc81114bfff3", :status ["done"]})

(doall (nrepl/message (nrepl/client conn 1000) {:op :eval :code "*ns*"}))
=>
({:id "71b1d920-63b8-46c3-b15b-6ed13dd9d21d",
  :ns "user",
  :session "ef99a436-0ab8-47b6-858a-13ad4077889e",
  :value "#object[clojure.lang.Namespace 0x262e3ef0 \"user\"]"}
 {:id "71b1d920-63b8-46c3-b15b-6ed13dd9d21d", :session "ef99a436-0ab8-47b6-858a-13ad4077889e", :status ["done"]})
```

It's possible Cursive had a bug previously accidentally implementing this behaviour, which is why it used to "work". Anyway, this issue can be closed.

**Comment by bbatsov** [comment]
Sorry for the radio silence here. If I recall correctly the responses were supposed to track the current ns (`*ns*`). If we just echo back the ns of the request that's not that useful IMO. I think last time I thought about this was at least 5 years ago. :D 

> Currently Cursive assumes that whatever comes back in a response message is the current ns, and that always used to work correctly. I don't know if this case previously didn't return :ns at all, or if it previously used to return the current value of *ns*. I'll see if I can test that today.

My memory of this is more or less the same, but I'll have to look into the code as well, and also see how I did the ns tracking in the REPL buffers in CIDER.

**Comment by cmf** [comment]
I'd be very interested to know what CIDER does to track this. I'll check older versions of Cursive to double check my memory of the behaviour there, but I'm 99.9% sure of it. 

I agree that sending `*ns*` back is not very useful when it's just the `:ns` that's passed in the message. Currently as far as I can tell it's actually impossible to affect `*ns*` from a message like this, presumably because when `*ns*` is updated it only affects the temporarily bound value. So it probably actually never makes sense to send `:ns` back when it's been sent in the request.

**Comment by shen-tian** [comment]
Just try `eval` on `(in-ns 'another.ns)`, and you'll get that ns back in the `:ns` slot:

```
(-->
  id                             "12"
  op                             "eval"
  session                        "0cb26984-ac15-443c-a7bd-72f0b66b69a1"
  time-stamp                     "2019-12-01 22:26:25.475589000"
  code                           "(in-ns 'another.ns)"
  column                         7
  content-type                   "true"
  file                           "*cider-repl Code/nrepl:localhost:61590(clj)*"
  line                           20
  nrepl.middleware.print/options (dict ...)
  nrepl.middleware.print/print   "cider.nrepl.pprint/pprint"
  nrepl.middleware.print/quota   1048576
  nrepl.middleware.print/stream? "1"
  ns                             "user"
)
(<--
  id         "12"
  session    "0cb26984-ac15-443c-a7bd-72f0b66b69a1"
  time-stamp "2019-12-01 22:26:25.516550000"
  value      "#namespace[another.ns]"
)
(<--
  id         "12"
  session    "0cb26984-ac15-443c-a7bd-72f0b66b69a1"
  time-stamp "2019-12-01 22:26:25.516942000"
  ns         "another.ns"
)
(<--
  id         "12"
  session    "0cb26984-ac15-443c-a7bd-72f0b66b69a1"
  time-stamp "2019-12-01 22:26:25.517085000"
  status     ("done")
)
```

**Comment by cmf** [comment]
@shen-tian Right, but if you send another evaluation after that evaluating `*ns*`, you'll get back whatever it was previously. The `(in-ns 'another.ns)` doesn't actually do anything at all - I believe this is because it affects the `*ns*` value which is temporarily bound when `:ns` is passed in the request (see my example [here](#issuecomment-560046372)).

**Comment by shen-tian** [comment]
Sorry, deleted that message there because I realised it was missing the point.

**Comment by shen-tian** [comment]
But as far as I can tell the return `:ns` value is just there so the client knows, during the course of the eval,  *ns* was changed, so it can update what the client's eval ns is? It's not there to confirm any persistent server side change. 

The design asks the client to keep track of current ns. The server doesn't maintain memory of it.

**Comment by cmf** [comment]
Well, the server does maintain memory of it, in `*ns*` - that's the source of truth for "the current namespace". And clients like CIDER and Cursive do have to track that, e.g. so that they can use the right namespace for autocompletion etc, and also show the user which namespace they're currently working in. While that's possible with the current nREPL protocol making the assumptions discussed above, it's decidedly non-obvious to someone implementing a client how this works. I think the current behaviour is unnecessarily confusing.

**Comment by shen-tian** [comment]
Agreed on the behaviour not being obvious. File under "need better documentation"?

**Comment by cmf** [comment]
I'd file under ''need better behaviour" and then document that :-)

**Comment by shen-tian** [comment]
Better behaviour would constitute a breaking change though?

**Comment by cmf** [comment]
Possibly, but it seems very unlikely that clients are relying on the current behaviour. But right, if they are required to use hacks to make things work, changing how this works might break those.

**Comment by shen-tian** [comment]
Just to check, I do think the current design is consistent in what it's trying to do:

- There are no ways to persistently change `*ns*` server side. But this really isn't an objective in and of itself.
- The server keeps the client updated about the effect of eval'd code on namespaces. If it doesn't the client can't know that if `foo` is eval'd after `(ns 'test2)`, the user meant `test2/foo`.
- The client is responsible for sending that ns back with subsequent requests in the same REPL window (v.s. a source file with its own namespace). This removes the need to actually change `*ns*`.

Between these, I'm sure clients have to rely on the current behaviour to work correctly.

**Comment by cmf** [comment]
* The client does have a way to change `*ns*` - they eval `in-ns` or something that desugars to it like `ns`. 
* Currently, the server doesn't keep the client updated correctly in the case where the client sends `:ns` in the request, since `in-ns` and friends don't actually affect `*ns*` in the way that nREPL reports.
* My understanding is that `*ns*` should not be sent on every evaluation, only when evaluation in some namespace other than `*ns*` is required. Cursive sends it when eval'ing a form from a file, for example, but not when entering a form in the REPL editor.

**Comment by bbatsov** [comment]
Btw, I checked how we do things in CIDER - basically we send the current ns everywhere and the REPL buffer sets as the current `ns` whatever it got from the last eval response that originated from it. Changing the response semantics won't break CIDER, and probably it won't break other clients either, but I guess first we should decide what's the optimal return value here - do we always return the current ns, do we return it only if it changed (this might be a bit confusing), do we introduce an extra key in the response.

**Comment by alexander-yakushev** [comment]
I went over this code recently; I didn't change the behavior but I'm confident I can speak how it works. Indeed, the `:ns` key that is returned with the eval result signifies which `*ns*` the evaluation finished with.
 
- If the evaluation didn't change `*ns*`, then `:ns` would be the same as the last one. 
- If the evaluation did change the `*ns*`, the `:ns` would be the new value.
- If the request provided an explicit `:ns`, then `:ns` in the response would be the same, but the `*ns*` variable will not be changed.

The client can use the above three facts to accurately know the state of `*ns*` after the evaluation. In any case, it is possible to query `*ns*` again if needed. @bbatsov, I think we can close this one as the behavior seems consistent to me.
